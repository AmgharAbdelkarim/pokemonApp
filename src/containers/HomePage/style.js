import styled from "styled-components";

export const ContentWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  background: #fff;
  .filter_box {
    width: 20%;
    border-right: 1px solid #e7e7e7;
    @media (max-width: 900px) {
      width: 100%;
      text-align: center;
      border-bottom: 1px solid #e7e7e7;
    }
    .clear_button {
      padding: 8px 10px;
      color: #007aff;
      font-size: 12px;
      letter-spacing: 0;
      line-height: 14px;
      &:hover {
        border-radius: 4px;
        background-color: #f7f7f7;
      }
    }
    > h5 {
      font-size: 18px;
      font-weight: bold;
      letter-spacing: 0;
      line-height: 23px;
      color: #04336b;
      margin: 8px 16px;
    }
    > ul {
      margin: 0px;
      list-style-type: none;
      padding: 0px 10px;

      .selected {
        border: 1px solid rgba(194, 209, 217, 0.27);
        border-radius: 4px;
        background-color: #f7f7f7;
      }
      > li {
        cursor: pointer;
        padding: 8px 12px;
        > span {
          color: #566684;
          font-size: 16px;
          font-weight: 500;
          padding: 8px;
        }
        &:hover {
          background: #e7e7e7;
        }
        margin-top: 12px;
      }
    }
  }
  .item_container {
    padding: 25px;
    width: 75%;
    @media (max-width: 900px) {
      width: 100%;
    }
    .item_wrapper {
      display: flex;
      justify-content: space-evenly;
      flex-wrap: wrap;
      margin-bottom: 30px;
    }
    .button_wrapper {
      display: flex;
      justify-content: center;
      margin: 20px 0;
    }
  }
`;
