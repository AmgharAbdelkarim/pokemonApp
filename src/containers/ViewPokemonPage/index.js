import React, { useEffect } from "react";

import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import Card from "components/Card";
import { PokemonContainer } from "containers/ViewPokemonPage/styles";
import Box from "components/Box";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonDetailsSelector,
  getLoadingSelector,
  getHasErrorSelector,
  getPokemonEvolutionSelector,
} from "store/selectors";
import { getPokemonDetails, getPokemonEvolutions } from "store/action";

const ViewPokemon = ({
  match,
  history,
  pokemon,
  getPokemonDetails,
  getPokemonEvolutions,
  evolution,
  loading,
  hasError,
}) => {
  const { params } = match;

  useEffect(() => {
    getPokemonDetails(params.name);
  }, [params.name]);
  useEffect(() => {
    getPokemonEvolutions(pokemon.id);
  }, [pokemon]);
  return (
    <PokemonContainer id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.name ? (
        <React.Fragment>
          <Card pokemon={pokemon} clickHandler={(url) => history.push(url)} />
          {evolution.length > 0 && (
            <React.Fragment>
              <h3>pokemon evolutions</h3>
              {evolution.map(({ species }, index) => (
                <Box
                  key={index}
                  name={species.name}
                  clickHandler={() => history.push(`/pokemon/${species.name}`)}
                />
              ))}
            </React.Fragment>
          )}
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </PokemonContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonDetailsSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
  evolution: getPokemonEvolutionSelector,
});

const mapDispatchToProps = {
  getPokemonDetails,
  getPokemonEvolutions,
};

export default connect(mapStateToProps, mapDispatchToProps)(ViewPokemon);
export { ViewPokemon };
