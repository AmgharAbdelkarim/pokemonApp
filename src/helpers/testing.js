import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import store from "store";
import { Router as Switch } from "react-router-dom";

import { createBrowserHistory } from "history";

export const history = createBrowserHistory();

export const Wrapper = (props) => {
  return (
    <Provider store={store()}>
      <Switch history={history}>{props.children}</Switch>
    </Provider>
  );
};

export const renderWithRedux = (ui, renderOptions = {}) => {
  return {
    ...render(ui, {
      wrapper: Wrapper,
      ...renderOptions,
    }),
    store,
  };
};
