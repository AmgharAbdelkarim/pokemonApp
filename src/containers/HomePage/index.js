import React, { useEffect } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { ContentWrapper } from "containers/HomePage/style";
import Pokemon from "components/Box";
import Button from "components/Button";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonSelector,
  getSizeSelector,
  getTypesSelector,
  getLoadingSelector,
  getHasErrorSelector,
  getCurrentUrlSelector,
  getNextUrlSelector,
  getPrevUrlSelector,
  getSelectSelectedTypes,
} from "store/selectors";
import {
  getPokemon,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
} from "store/action";

const Home = ({
  pokemon,
  size,
  types,
  loading,
  hasError,
  currentUrl,
  nextUrl,
  prevUrl,
  getPokemon,
  selectedTypes,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
  history,
}) => {
  useEffect(() => {
    getPokemon();
  }, [currentUrl]);

  useEffect(() => {
    getTypes();
  }, []);

  const nextHandler = () => {
    if (!nextUrl) return setSize(size + 20);
    setCurrentUrl(nextUrl);
  };

  const prevHandler = () => {
    if (!nextUrl) return setSize(size - 20);
    setCurrentUrl(prevUrl);
  };

  const filterByType = (currentUrl, name) => {
    setCurrentUrl(currentUrl);
    setTypeSelected(name);
  };

  return (
    <ContentWrapper id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.length > 0 ? (
        <React.Fragment>
          <div className="filter_box">
            <h5>filter Pokemon By Type</h5>
            {selectedTypes.length > 0 && (
              <Button
                className="clear_button"
                clickHandler={() => {
                  setCurrentUrl("https://pokeapi.co/api/v2/pokemon");
                  clearTypeSelected();
                }}
              >
                Clear
              </Button>
            )}
            <ul>
              {types.map(({ name, url }) => (
                <li
                  onClick={() => filterByType(url, name)}
                  aria-hidden="true"
                  key={name}
                  className={selectedTypes.indexOf(name) >= 0 ? "selected" : ""}
                >
                  <span>{name}</span>
                </li>
              ))}
            </ul>
          </div>
          <div className="item_container">
            <div className="item_wrapper">
              {pokemon.slice(size, size + 20).map((p, index) => (
                <Pokemon
                  key={index}
                  name={p.name || p.pokemon.name}
                  clickHandler={() =>
                    history.push(`/pokemon/${p.name || p.pokemon.name}`)
                  }
                />
              ))}
            </div>
            <div className="button_wrapper">
              <Button
                disabled={prevUrl || size > 19 ? false : true}
                clickHandler={() => prevHandler()}
              >
                previous
              </Button>
              <Button
                disabled={nextUrl || size < pokemon.length - 19 ? false : true}
                clickHandler={() => nextHandler()}
              >
                next
              </Button>
            </div>
          </div>
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </ContentWrapper>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonSelector,
  size: getSizeSelector,
  types: getTypesSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
  currentUrl: getCurrentUrlSelector,
  nextUrl: getNextUrlSelector,
  prevUrl: getPrevUrlSelector,
  selectedTypes: getSelectSelectedTypes,
});

const mapDispatchToProps = {
  getPokemon,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
};
export default connect(mapStateToProps, mapDispatchToProps)(Home);

export { Home };
