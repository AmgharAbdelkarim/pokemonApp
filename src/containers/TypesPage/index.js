import React, { useEffect } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import Pokemon from "components/Box";
import Button from "components/Button";
import { TypeWrapper } from "containers/TypesPage/style";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonTypesSelector,
  getSizeSelector,
  getLoadingSelector,
  getHasErrorSelector,
} from "store/selectors";
import { getPokemonTypes, setSize } from "store/action";

const Types = ({
  match,
  pokemon,
  size,
  loading,
  hasError,
  getPokemonTypes,
  setSize,
  history,
}) => {
  const { params } = match;

  useEffect(() => {
    getPokemonTypes(params.type);
  }, []);
  const nextHandler = () => {
    setSize(size + 20);
  };
  const prevHandler = () => {
    setSize(size - 20);
  };
  return (
    <TypeWrapper id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.length > 0 ? (
        <React.Fragment>
          <div className="item_wrapper">
            {pokemon.slice(size, size + 20).map(({ pokemon }, index) => (
              <Pokemon
                clickHandler={() => history.push(`/pokemon/${pokemon.name}`)}
                name={pokemon.name}
                key={index}
              />
            ))}
          </div>
          <div className="button_wrapper">
            <Button
              disabled={size > 19 ? false : true}
              clickHandler={() => prevHandler()}
            >
              previous
            </Button>
            <Button
              disabled={size < pokemon.length - 19 ? false : true}
              clickHandler={() => nextHandler()}
            >
              next
            </Button>
          </div>
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </TypeWrapper>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonTypesSelector,
  size: getSizeSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
});

const mapDispatchToProps = {
  getPokemonTypes,
  setSize,
};

export default connect(mapStateToProps, mapDispatchToProps)(Types);

export { Types };
