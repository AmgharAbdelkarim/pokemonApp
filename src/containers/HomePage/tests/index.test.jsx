import { cleanup, fireEvent, waitFor } from "@testing-library/react";

import { renderWithRedux } from "helpers/testing";

import { Home } from "containers/HomePage";
import { createBrowserHistory } from "history";

afterEach(cleanup);

describe("Home", () => {
  let props;
  beforeEach(() => {
    props = {
      pokemon: [
        {
          name: "bulbasaur",
          url: "https://pokeapi.co/api/v2/pokemon/1/",
        },
        {
          name: "ivysaur",
          url: "https://pokeapi.co/api/v2/pokemon/2/",
        },
        {
          name: "venusaur",
          url: "https://pokeapi.co/api/v2/pokemon/3/",
        },
        {
          name: "charmander",
          url: "https://pokeapi.co/api/v2/pokemon/4/",
        },
        {
          name: "charmeleon",
          url: "https://pokeapi.co/api/v2/pokemon/5/",
        },
        {
          name: "charizard",
          url: "https://pokeapi.co/api/v2/pokemon/6/",
        },
        {
          name: "squirtle",
          url: "https://pokeapi.co/api/v2/pokemon/7/",
        },
        {
          name: "wartortle",
          url: "https://pokeapi.co/api/v2/pokemon/8/",
        },
        {
          name: "blastoise",
          url: "https://pokeapi.co/api/v2/pokemon/9/",
        },
        {
          name: "caterpie",
          url: "https://pokeapi.co/api/v2/pokemon/10/",
        },
        {
          name: "metapod",
          url: "https://pokeapi.co/api/v2/pokemon/11/",
        },
        {
          name: "butterfree",
          url: "https://pokeapi.co/api/v2/pokemon/12/",
        },
        {
          name: "weedle",
          url: "https://pokeapi.co/api/v2/pokemon/13/",
        },
        {
          name: "kakuna",
          url: "https://pokeapi.co/api/v2/pokemon/14/",
        },
        {
          name: "beedrill",
          url: "https://pokeapi.co/api/v2/pokemon/15/",
        },
        {
          name: "pidgey",
          url: "https://pokeapi.co/api/v2/pokemon/16/",
        },
        {
          name: "pidgeotto",
          url: "https://pokeapi.co/api/v2/pokemon/17/",
        },
        {
          name: "pidgeot",
          url: "https://pokeapi.co/api/v2/pokemon/18/",
        },
        {
          name: "rattata",
          url: "https://pokeapi.co/api/v2/pokemon/19/",
        },
        {
          name: "raticate",
          url: "https://pokeapi.co/api/v2/pokemon/20/",
        },
      ],
      size: 0,
      types: [],
      loading: false,
      hasError: false,
      currentUrl: "https://pokeapi.co/api/v2/pokemon",
      nextUrl: "https://pokeapi.co/api/v2/pokemon?offset=20&limit=20",
      prevUrl: "https://pokeapi.co/api/v2/pokemon",
      getPokemon: jest.fn(),
      selectedTypes: [],
      setTypeSelected: jest.fn(),
      setCurrentUrl: jest.fn(),
      setSize: jest.fn(),
      getTypes: jest.fn(),
      clearTypeSelected: jest.fn(),
    };
  });
  afterEach(cleanup);
  it("should Home render correctly", () => {
    const { asFragment } = renderWithRedux(<Home {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should render loading state if loading is true", () => {
    let newProps = { ...props, loading: true };
    const { getByText } = renderWithRedux(<Home {...newProps} />);
    expect(getByText("Loading...")).toBeInTheDocument();
  });
  it("should render error state if hasError is true", () => {
    let newProps = { ...props, hasError: true };
    const { getByText } = renderWithRedux(<Home {...newProps} />);
    expect(getByText("Something Went Wrong")).toBeInTheDocument();
  });
  it("should call setCurrentUrl action when we click on nextButton", async () => {
    const { getByRole } = renderWithRedux(<Home {...props} />);
    const nextButton = getByRole("button", { name: /next/ });
    fireEvent.click(nextButton);
    await waitFor(() =>
      expect(props.setCurrentUrl).toHaveBeenCalledWith(props.nextUrl)
    );
  });
  it("should call setCurrentUrl action when we click on prevButton", async () => {
    const { getByRole } = renderWithRedux(<Home {...props} />);
    const prevButton = getByRole("button", { name: /previous/ });
    fireEvent.click(prevButton);
    await waitFor(() =>
      expect(props.setCurrentUrl).toHaveBeenCalledWith(props.prevUrl)
    );
  });
  it("should next button be disabled if next url is empty", () => {
    let newProps = { ...props, size: 20, nextUrl: "" };
    const { getByRole } = renderWithRedux(<Home {...newProps} />);
    const nextButton = getByRole("button", { name: /next/ });
    expect(nextButton).toBeDisabled();
  });
  it("should next button be disabled if prev url is empty", async () => {
    let newProps = { ...props, prevUrl: "" };
    const { getByRole } = renderWithRedux(<Home {...newProps} />);
    const prevButton = getByRole("button", { name: /previous/ });
    expect(prevButton).toBeDisabled();
  });

  it("should redirect to view pokemon page if i click on pokemon", async () => {
    const history = createBrowserHistory();
    jest.spyOn(history, "push");
    const { getByText } = renderWithRedux(
      <Home {...props} history={history} />
    );
    const pokemon = getByText("bulbasaur");
    fireEvent.click(pokemon);
    await waitFor(() =>
      expect(history.push).toHaveBeenCalledWith("/pokemon/bulbasaur")
    );
  });
});
