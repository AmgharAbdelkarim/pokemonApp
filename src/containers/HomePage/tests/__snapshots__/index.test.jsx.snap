// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should Home render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ kbYcHs"
    id="content"
  >
    <div
      class="filter_box"
    >
      <h5>
        filter Pokemon By Type
      </h5>
      <ul />
    </div>
    <div
      class="item_container"
    >
      <div
        class="item_wrapper"
      >
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            bulbasaur
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            ivysaur
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            venusaur
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            charmander
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            charmeleon
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            charizard
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            squirtle
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            wartortle
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            blastoise
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            caterpie
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            metapod
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            butterfree
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            weedle
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            kakuna
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            beedrill
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            pidgey
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            pidgeotto
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            pidgeot
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            rattata
          </span>
        </div>
        <div
          class="sc-gsTCUz dfCzHG"
        >
          <span>
            raticate
          </span>
        </div>
      </div>
      <div
        class="button_wrapper"
      >
        <div
          class="sc-dlfnbm eciQpr"
        >
          <button>
            previous
          </button>
        </div>
        <div
          class="sc-dlfnbm eciQpr"
        >
          <button>
            next
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
